staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugins
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/klog"
+
+	v1 "k8s.io/api/core/v1"
+	storage "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// VSphereDriverName is the name of the CSI driver for vSphere Volume
+	VSphereDriverName = "csi.vsphere.vmware.com"
+	// VSphereInTreePluginName is the name of the in-tree plugin for vSphere Volume
+	VSphereInTreePluginName = "kubernetes.io/vsphere-volume"
+	// Name of parameter to supply SPBM Policy name for Volume Provisioning
+	paramStoragePolicyName = "storagepolicyname"
+	// Name of parameter to let controller know that it needs to return volumefilePath as Volume Attribute
+	// When CSI Driver Provision Volume using in-tree Storage Class and PVC
+	paramRequestVolumeFilePath   = "requestvolumefilepath"
+	paramVolumeFilePathAttribute = "volumefilePath"
+)
+
+var _ InTreePlugin = &vSphereCSITranslator{}
+
+// vSphereCSITranslator handles translation of PV spec from In-tree vSphere Volume to vSphere CSI
+type vSphereCSITranslator struct{}
+
+// NewvSphereCSITranslator returns a new instance of vSphereCSITranslator
+func NewvSphereCSITranslator() InTreePlugin {
+	return &vSphereCSITranslator{}
+}
+
+// TranslateInTreeStorageClassToCSI translates InTree vSphere storage class parameters to CSI storage class
+func (t *vSphereCSITranslator) TranslateInTreeStorageClassToCSI(sc *storage.StorageClass) (*storage.StorageClass, error) {
+	if sc == nil {
+		return nil, fmt.Errorf("sc is nil")
+	}
+	var params = map[string]string{}
+	// only "storagepolicyname" SC parameter in vsphere in-tree volume is
```

Comment:
yes, @misterikkit inline vSAN raw policy parameter is no longer supported in vSphere CNS. 

so the following sc parameters can not be supported after migration.

- hostfailurestotolerate
- forceprovisioning
- cachereservation
- diskstripes
- objectspacereservation
- iopslimit

vSphere CNS helps manages the disk's reservation using vSAN SPBM policy setting - `Object space reservation`. so we no longer need to supply this from storage class in the CSI driver.

datastore name is not supported in the vSphere CSI driver and new vSphere CNS solution. The reason is name is not durable. so we have introduced datastoreURL instead. Considering the use cases of this parameter, we can achieve the same with the SPBM policy if configured properly for datastore selection, so no plan for migration support for this parameter as well. 

Existing provisioned volume using these parameters will be migrated and will be handled by CSI, but new volumes creation using in-tree storage class using these parameters will not be honored by the vSphere CSI driver.

Reply:
It would be good to document the unsupported parameters and their new equivalents somewhere, in the driver maybe?

Reply:
yes we will publish migration documentation at - https://vsphere-csi-driver.sigs.k8s.io/

will list all deprecated parameters there.

Reply:
@divyenpatel If I understand, datastore name might change, but datastore UID cannot change. And if an admin changed datastore name, they would break existing storage classes and PVs that used the in-tree plugin.

Another difference I noticed between the in-tree vSphere plugin and the vSphere CSI driver is that the PersistentVolumes reference VMDKs by UID. This means that the datastore name can change, the VMDK name can change, and the VMDK can move to a different datastore, all without breaking k8s.

However, for the purposes of migration, this implementation passes the legacy API `volumePath` field as the volume handle, requiring the CSI driver to resolve datastore and VMDK by name. Could we do something similar for storage classes that is only available as part of migration, and not as a regular feature of the CSI driver?

Reply:
in-tree VCP uses following as volume path

```
[vsanDatastore] 6785a85e-268e-6352-a2e8-02008b7afadd/kubernetes-dynamic-pvc-68734c9f-a679-42e6-a694-39632c51e31f.vmdk
```

here `vsanDatastore` is the name of the datastore. if this name changed after volume is provisioned. volume can not be used unless we patch or re-create PV with a new datastore name in the path. This was a major limitation on in-tree volume. This is fixed with the vSphere CSI driver. vSphere CSI driver used First Class Disk ID as volume handle using which disk can be directly managed by vCenter APIs. you can change the datastore name with vSphere CSI driver.

> Could we do something similar for storage classes that is only available as part of migration

This will require a vSphere CSI driver to relax SC parameter validation to allow passing datastorename.  Existing volumes created with `datastoename` parameter supplied in the SC, will get migrated, but why do user need to create a new volumes with older in-tree YAML when the CSI driver is already installed on the cluster?

Reply:
> This was a major limitation on in-tree volume.

Yes! The CSI driver solves this problem. In the context of CSI migration, we *do not need* to solve this problem for in-tree VCP. It is okay & expected for _migrated_ volumes to have this problem. Also, this problem seems to apply to both volume create and volume attach.



> why do user need to create a new volumes with older in-tree YAML when the CSI driver is already installed on the cluster?

This is one of the core assumptions of CSI migration. We can never remove the legacy API (until k8s v2) and we cannot enforce all users switching to CSI. If we didn't need to support volume creation, then we would not translate storage class at all.

> This will require a vSphere CSI driver to relax SC parameter validation

It seems that you added a new storage class parameter, `requestvolumefilepath`, to the CSI driver, whose only purpose is to populate `volumePath` in a legacy PV during volume creation. But a user would never see this parameter because it is only set by the translation library. Are you comfortable having a similar, "migration only" parameter for datastore name?

Reply:
> But a user would never see this parameter because it is only set by the translation library. Are you comfortable having a similar, "migration only" parameter for datastore name?

parameters can not stay migration only.

This will require the vSphere CSI driver to accept `datastore` as a valid parameter. vSphere CSI driver users can also use this parameter by specifying it in the storage class.
Adding support for` datastore` does not align with our plan.

Reply:
@divyenpatel As we discussed lets encode all the old params into a new storage class param(say `csiMigrationParams`) during translation. This way we can decide which old params to support in CSI based on customer feedback.

Reply:
supplying all param in JSON encoded string as the value for `csiMigrationParams`.